File: src/bus_rl/sim/passengers.py
"""FIFO passenger boarding, alighting, and abandonment."""

from __future__ import annotations

from dataclasses import dataclass

from bus_rl.domain import PassengerCohort, PassengerStatus, Pattern, Vehicle, WorldState


@dataclass(frozen=True)
class PassengerEvents:
    boarded_count: int = 0
    alighted_count: int = 0
    first_denied_count: int = 0


def _eligible(cohort: PassengerCohort, bus: Vehicle, direction: int, stop_index: int) -> bool:
    if (
        cohort.status is not PassengerStatus.WAITING
        or cohort.direction != direction
        or cohort.origin_index != stop_index
    ):
        return False
    if bus.pattern is Pattern.FULL:
        return True
    turn = bus.turn_stop if bus.turn_stop is not None else 3
    if direction == 1:
        return cohort.destination_index <= turn
    return stop_index <= turn


def _split_for_boarding(
    state: WorldState, cohort: PassengerCohort, count: int, bus_id: int
) -> None:
    boarded = PassengerCohort(
        cohort_id=state.next_cohort_id,
        lineage_id=cohort.lineage_id,
        route_id=cohort.route_id,
        direction=cohort.direction,
        origin_index=cohort.origin_index,
        destination_index=cohort.destination_index,
        arrival_tick=cohort.arrival_tick,
        count=count,
        status=PassengerStatus.ONBOARD,
        first_denied=cohort.first_denied,
        boarding_tick=state.current_time_s // 30,
    )
    state.next_cohort_id += 1
    cohort.count -= count
    state.cohorts.append(boarded)
    state.vehicles[bus_id].passengers.append(boarded)


def board_visit(
    state: WorldState, bus_id: int, route_id: int, direction: int, stop_index: int
) -> PassengerEvents:
    """Board FIFO eligible cohorts at a real visit; never moves the vehicle itself."""
    bus = state.vehicles[bus_id]
    if bus.route_id != route_id:
        raise ValueError("bus route does not match visit route")
    if bus.load > bus.capacity:
        raise AssertionError("bus is already over capacity")
    candidates = sorted(
        (
            c
            for c in state.cohorts
            if c.route_id == route_id and _eligible(c, bus, direction, stop_index)
        ),
        key=lambda c: (c.arrival_tick, c.cohort_id),
    )
    boarded = denied = 0
    for cohort in candidates:
        available = bus.capacity - bus.load
        take = min(available, cohort.count)
        if take:
            _split_for_boarding(state, cohort, take, bus_id)
            boarded += take
        if cohort.count and not cohort.first_denied:
            cohort.first_denied = True
            denied += cohort.count
    state.cohorts[:] = [c for c in state.cohorts if c.count > 0]
    state.assert_conservation()
    return PassengerEvents(boarded_count=boarded, first_denied_count=denied)


def alight_visit(state: WorldState, bus_id: int, stop_index: int) -> int:
    bus = state.vehicles[bus_id]
    alighted = 0
    for cohort in bus.passengers:
        if cohort.destination_index == stop_index and cohort.status is PassengerStatus.ONBOARD:
            cohort.status = PassengerStatus.COMPLETED
            cohort.completion_tick = state.current_time_s // 30
            alighted += cohort.count
    bus.passengers[:] = [c for c in bus.passengers if c.status is PassengerStatus.ONBOARD]
    state.assert_conservation()
    return alighted


def abandon_expired(state: WorldState, patience_s: int, tick_s: int) -> int:
    abandoned = 0
    for cohort in state.cohorts:
        if (
            cohort.status is PassengerStatus.WAITING
            and state.current_time_s - cohort.arrival_tick * tick_s >= patience_s
        ):
            cohort.status = PassengerStatus.ABANDONED
            cohort.abandonment_tick = state.current_time_s // tick_s
            abandoned += cohort.count
    state.assert_conservation()
    return abandoned

File: crates/bus-sim/src/passengers.rs
//! FIFO passenger boarding, alighting, and abandonment.
//!
//! Port of `src/bus_rl/sim/passengers.py`. Tick stamps use the literal `30`
//! exactly like the oracle (discrepancy D4); only `tick_s=30` is supported.

use crate::domain::{PassengerCohort, PassengerStatus, Pattern, WorldState};

/// Boarding/completion/abandonment tick stamp divisor (oracle discrepancy D4).
const TICK_STAMP_S: i64 = 30;

#[derive(Clone, Copy, Debug, Default)]
pub struct PassengerEvents {
    pub boarded_count: i64,
    pub first_denied_count: i64,
}

fn eligible(
    cohort: &PassengerCohort,
    pattern: Pattern,
    turn_stop: u32,
    direction: i32,
    stop_index: u32,
) -> bool {
    if cohort.status != PassengerStatus::Waiting
        || cohort.direction != direction
        || cohort.origin_index != stop_index
    {
        return false;
    }
    if pattern == Pattern::Full {
        return true;
    }
    if direction == 1 {
        cohort.destination_index <= turn_stop
    } else {
        stop_index <= turn_stop
    }
}

fn split_for_boarding(state: &mut WorldState, index: usize, count: i64, bus_id: usize) {
    let source = state.cohorts[index];
    let boarded = PassengerCohort {
        cohort_id: state.next_cohort_id,
        lineage_id: source.lineage_id,
        route_id: source.route_id,
        direction: source.direction,
        origin_index: source.origin_index,
        destination_index: source.destination_index,
        arrival_tick: source.arrival_tick,
        count,
        status: PassengerStatus::Onboard,
        first_denied: source.first_denied,
        boarding_tick: Some(state.current_time_s / TICK_STAMP_S),
        completion_tick: None,
        abandonment_tick: None,
    };
    state.next_cohort_id += 1;
    state.cohorts[index].count -= count;
    state.waiting_total -= count;
    state.onboard_total += count;
    let bus = &mut state.vehicles[bus_id];
    bus.passengers.push(boarded);
    bus.load += count;
}

/// Board FIFO eligible cohorts at a real visit; never moves the vehicle itself.
pub fn board_visit(
    state: &mut WorldState,
    bus_id: usize,
    route_id: u32,
    direction: i32,
    stop_index: u32,
) -> PassengerEvents {
    let (pattern, turn_stop, capacity, load, bus_route) = {
        let bus = &state.vehicles[bus_id];
        (
            bus.pattern,
            bus.turn_stop.unwrap_or(3),
            bus.capacity,
            bus.load,
            bus.route_id,
        )
    };
    assert_eq!(
        bus_route,
        Some(route_id),
        "bus route does not match visit route"
    );
    assert!(load <= capacity, "bus is already over capacity");

    let mut candidates: Vec<(i64, i64, usize)> = state
        .cohorts
        .iter()
        .enumerate()
        .filter(|(_, cohort)| eligible(cohort, pattern, turn_stop, direction, stop_index))
        .filter(|(_, cohort)| cohort.route_id == route_id)
        .map(|(index, cohort)| (cohort.arrival_tick, cohort.cohort_id, index))
        .collect();
    candidates.sort_unstable();

    let mut boarded = 0i64;
    let mut denied = 0i64;
    for (_, _, index) in candidates {
        let available = capacity - state.vehicles[bus_id].load;
        let count = state.cohorts[index].count;
        let take = available.min(count);
        if take > 0 {
            split_for_boarding(state, index, take, bus_id);
            boarded += take;
        }
        if state.cohorts[index].count > 0 && !state.cohorts[index].first_denied {
            state.cohorts[index].first_denied = true;
            denied += state.cohorts[index].count;
        }
    }
    state.cohorts.retain(|cohort| cohort.count > 0);
    PassengerEvents {
        boarded_count: boarded,
        first_denied_count: denied,
    }
}

pub fn alight_visit(state: &mut WorldState, bus_id: usize, stop_index: u32) -> i64 {
    let mut alighted = 0i64;
    let mut remaining: Vec<PassengerCohort> = Vec::new();
    let passengers = std::mem::take(&mut state.vehicles[bus_id].passengers);
    for mut cohort in passengers {
        if cohort.destination_index == stop_index && cohort.status == PassengerStatus::Onboard {
            cohort.status = PassengerStatus::Completed;
            cohort.completion_tick = Some(state.current_time_s / TICK_STAMP_S);
            alighted += cohort.count;
            state.onboard_total -= cohort.count;
            state.completed_total += cohort.count;
            state.finished.push(cohort);
            state.push_recent_finished(cohort);
        } else {
            remaining.push(cohort);
        }
    }
    state.vehicles[bus_id].passengers = remaining;
    state.vehicles[bus_id].load -= alighted;
    alighted
}

pub fn abandon_expired(state: &mut WorldState, patience_s: i64, tick_s: i64) -> i64 {
    let mut abandoned = 0i64;
    let mut remaining: Vec<PassengerCohort> = Vec::with_capacity(state.cohorts.len());
    let cohorts = std::mem::take(&mut state.cohorts);
    for mut cohort in cohorts {
        if state.current_time_s - cohort.arrival_tick * tick_s >= patience_s {
            cohort.status = PassengerStatus::Abandoned;
            cohort.abandonment_tick = Some(state.current_time_s / TICK_STAMP_S);
            abandoned += cohort.count;
            state.waiting_total -= cohort.count;
            state.abandoned_total += cohort.count;
            state.finished.push(cohort);
            state.push_recent_finished(cohort);
        } else {
            remaining.push(cohort);
        }
    }
    state.cohorts = remaining;
    abandoned
}
